docs/_design/map-controller.md

```diff
@@ -0,0 +1,399 @@
+1. It's unclear what should happen if multiple decorator objects match a single
+   parent object.
+   We could send multiple decorator objects to the hook, but that just passes
+   the complexity on to the user.
+1. It's unclear whether decorator objects are expected to take part in ownership
+   of the objects created.
+   Depending on the use case, users might want attachments to be owned by just
+   the parent, just the decorator, or both.
+   This configuration adds to the cognitive overhead of using the API,
+   and there's no one default that's more intuitive than the others.
+
+## Example
+
+The example use case we'll consider in this doc is a controller called
+SnapshotSchedule that creates periodic backups of PVCs with the VolumeSnapshot
+API.
+Notice that it's natural to express this in the form we defined above:
+"For every matching PVC, I want to create some VolumeSnapshot objects."
+
+CompositeController doesn't fit this use case because the PVCs are created and
+potentially owned by something other than the SnapshotSchedule object.
+For example, the PVCs might have been created by a StatefulSet.
+Instead of creating PVCs, we want to look at all the PVCs that already exist and
+take action on certain ones.
+
+DecoratorController doesn't fit this use case because it doesn't make sense for
+the VolumeSnapshots we create to be owned by the PVC from which the snapshot was
+taken.
+The lifecycle of a VolumeSnapshot has to be separate from the PVC because the
+whole point is that you should be able to recover the data if the PVC goes away.
+Since the PVC doesn't own the VolumeSnapshots, it doesn't make sense to think of
+the snapshots as a decoration on PVC (an additional feature of the PVC API).
+
+An instance of SnapshotSchedule might look like this:
+
+```yaml
+apiVersion: snapshot.k8s.io/v1
+kind: SnapshotSchedule
+metadata:
+  name: my-app-snapshots
+spec:
+  snapshotInterval: 6h
+  snapshotTTL: 10d
+  selector:
+    matchLabels:
+      app: my-app
+```
+
+It contains a selector that determines which PVCs this schedule applies to,
+and some parameters that determine how often to take snapshots, as well as when
+to retire old snapshots.
+
+## API
+
+Below is a sample MapController spec that could be used to implement the
+SnapshotSchedule controller:
+
+```yaml
+apiVersion: metacontroller.k8s.io/v1alpha1
+kind: MapController
+metadata:
+  name: snapshotschedule-controller
+spec:
+  parentResource:
+    apiVersion: snapshot.k8s.io/v1
+    resource: snapshotschedules
+  inputResources:
+  - apiVersion: v1
+    resource: persistentvolumeclaims
+  outputResources:
+  - apiVersion: volumesnapshot.external-storage.k8s.io/v1
+    resource: volumesnapshots
+  resyncPeriodSeconds: 5
+  hooks:
+    map:
+      webhook:
+        url: http://snapshotschedule-controller.metacontroller/map
+    tombstone:
+      webhook:
+        url: http://snapshotschedule-controller.metacontroller/tombstone
+```
```

Comment:
IIUC, parent owns outputs (label/selector + ownerRef), and selects inputs (label/selecto only)?

Can a user make map controller enter crazy creation loop by having the same resource in input and output?

Reply:
Is there any example of use cases involving same type of input and output resources? I think it will be a very infrequent use case for `MapController`. IIUC, both input and output resources will be different even when they are of same type (because the output resources are new resources created by the `MapController`). Hence, I think this use case may be better served by other meta-controllers.

Reply:
As synced by person, there are some use cases involving same type of input and output resources like:
1. pod as input, another pod as output
2. service as input, companion service as output
3. ConfigMap with data A as input, ConfigMap with data B as output

Reply:
> Since we put ownerRef on the outputs like you mentioned, we could ignore our own outputs when looking at inputs. 

This stops resources from using their own children as inputs, which seems fine. 

> We'll never prevent all possible hot loops (you can set up multiple controllers, for example) so I just see our responsibility as providing reasonable guardrails against common mistakes.

Agree. Adding a note/caveat is probably enough.